File: 4Periodo/EstruturasDeDados/2Avaliacao/Atividades/5Encontro/lista_encadeada.go
package encontro

import "errors"

type Node struct {
	dado    int
	proximo *Node
}
type ListaEncadeada struct {
	inicio *Node
	fim    *Node
}

func (l *ListaEncadeada) inserir_inicio(dado int) {
	var primeiro *Node
	primeiro.dado = dado
	primeiro.proximo = l.inicio
	l.inicio = primeiro
}

func (l *ListaEncadeada) inserir_fim(dado int) {
	if l == nil {
		return
	}
	novo := &Node{dado: dado, proximo: nil}
	if l.fim == nil {
		l.inicio = novo
		l.fim = novo
		return
	}

	l.fim.proximo = novo
	l.fim = novo
}

var (
	ErrListNotFound    = errors.New("linked list is nil or empty")
	ErrElementNotFound = errors.New("element was not found")
)

func (l *ListaEncadeada) remover(valor int) error {
	if l == nil || l.inicio == nil {
		return ErrListNotFound
	}
	atual := l.inicio
	var anterior *Node
	for atual != nil && atual.dado != valor {
		anterior = atual
		atual = atual.proximo
	}
	if atual == nil {
		return ErrElementNotFound
	}
	if anterior == nil {
		l.inicio = atual.proximo
	} else {
		anterior.proximo = atual.proximo
	}
	if atual == l.fim {
		l.fim = anterior
	}
	return nil
}

File: 4Periodo/EstruturasDeDados/2Avaliacao/Atividades/5Encontro/LinguagemC/lista_encadeada.c
// go:build ignore

#include <stdlib.h>
typedef struct No {
  int dado;
  struct No *prox;
} No;
typedef struct ListaEncadeada {
  No *inicio;
} Lista;
void inserir_inicio(Lista *l, int dado) {
  No *primeiro = NULL;
  primeiro = malloc(sizeof(No));
  if (primeiro == NULL)
    return;
  primeiro->dado = dado;
  primeiro->prox = l->inicio;
  l->inicio = primeiro;
}
void inserir_fim(Lista *l, int dado) {
  if (l == NULL) {
    return;
  }
  No *novo = malloc(sizeof(No));
  if (novo == NULL)
    return;
  novo->dado = dado;
  novo->prox = NULL;

  if (l->inicio == NULL) {
    l->inicio = novo;
    return;
  }
  No *atual = l->inicio;
  while (atual->prox != NULL) {
    atual = atual->prox;
  }
  atual->prox = novo;
}
int remover(Lista *l, int valor) {
  if (l == NULL || l->inicio == NULL) {
    return 0;
  }
  No *atual = l->inicio;
  No *anterior = NULL;

  while (atual != NULL && atual->dado != valor) {
    anterior = atual;
    atual = atual->prox;
  }
  if (atual == NULL) {
    return 0;
  }
  if (anterior == NULL) {
    l->inicio = atual->prox;
  } else {
    anterior->prox = atual->prox;
  }
  free(atual);
  return 1;
}

void liberar(Lista *l) {
  if (l == NULL)
    return;
  No *node_antigo = l->inicio;
  while (node_antigo != NULL) {
    No *node_novo = node_antigo->prox;
    free(node_antigo);
    node_antigo = node_novo;
  }
  l->inicio = NULL;
}
